lib/starttls.js

```diff
@@ -33,111 +15,46 @@ function starttlsClient(socket, options, callback) {
 }
 
 function starttls(socket, options, callback, isServer) {
-  var sslcontext;
-
   var opts = {};
+
   Object.keys(options).forEach(function(key) {
     opts[key] = options[key];
   });
   if (!opts.ciphers) {
     opts.ciphers = RECOMMENDED_CIPHERS;
   }
+  opts.isServer = isServer;
+  opts.secureContext = tls.createSecureContext(opts);
 
-  socket.removeAllListeners('data');
-  if (tls.createSecureContext) {
-    sslcontext = tls.createSecureContext(opts);
-  } else {
-    sslcontext = crypto.createCredentials(opts);
-  }
-  var pair = tls.createSecurePair(sslcontext, isServer);
-  var cleartext = pipe(pair, socket);
-
+  var secureSocket = new tls.TLSSocket(socket, opts);
   var erroredOut = false;
-  pair.on('secure', function() {
+
+  // NodeJS documentation bug: secure vs secureConnect
+  // https://github.com/nodejs/node/issues/10555
+  secureSocket.on('secure', function() {
     if (erroredOut) {
-      pair.end();
+      secureSocket.end();
       return;
     }
 
-    var verifyError = (pair._ssl || pair.ssl).verifyError();
-
-    if (verifyError) {
-      cleartext.authorized = false;
-      cleartext.authorizationError = verifyError;
+    var authError = secureSocket.ssl.verifyError();
+    if (authError) {
+        secureSocket.authorized = false;
+        secureSocket.authorizationError = authError;
     } else {
-      cleartext.authorized = true;
+        secureSocket.authorized = true;
```

Comment:
> when TLS handshake happens on control connection we don't even know client's login credentials

Right, so we'd have to save which name the cert claimed to certify (if we believe it) so we can remember it when a USER or PASS command arrives.